src/moonlink/src/storage/mooncake_table.rs

```diff
@@ -754,84 +763,172 @@ impl MooncakeTable {
         }
+        self.next_snapshot_task.new_mem_indices.push(index.clone());
 
-        let path = metadata.path.clone();
-        let parquet_flush_threshold_size = metadata.config.disk_slice_parquet_file_size;
+        let path = self.metadata.path.clone();
+        let parquet_flush_threshold_size = self.metadata.config.disk_slice_parquet_file_size;
+        let next_file_id = self.next_file_id;
+        self.next_file_id += 1;
 
-        let mut disk_slice = DiskSliceWriter::new(
-            metadata.schema.clone(),
+        let disk_slice = DiskSliceWriter::new(
+            self.metadata.schema.clone(),
             path,
             batches,
-            lsn,
+            Some(lsn),
             next_file_id,
             index,
             parquet_flush_threshold_size,
         );
 
-        disk_slice.write().await?;
         Ok(disk_slice)
     }
 
-    // UNDONE(BATCH_INSERT):
-    // Flush uncommitted batches from big batch insert, whether how much record batch there is.
-    //
-    // This function
-    // - tracks all record batches by current snapshot task
-    // - persists all full batch records to local filesystem
-    pub async fn flush(&mut self, lsn: u64) -> Result<()> {
-        // Sanith check flush LSN doesn't regress.
+    /// Flushes the disk slice for the transaction.
+    fn flush_disk_slice(
+        &mut self,
+        disk_slice: &mut DiskSliceWriter,
+        table_notify_tx: Sender<TableEvent>,
+        xact_id: Option<u32>,
+    ) {
+        let path = self.metadata.path.clone();
+        let schema = self.metadata.schema.clone();
+        let batches = disk_slice.take_input_batches();
+        let lsn = disk_slice.lsn();
+        let next_file_id = disk_slice.table_auto_incr_id;
+        let old_index = disk_slice.old_index().clone();
+        let parquet_flush_threshold_size = self.metadata.config.disk_slice_parquet_file_size;
+
+        if let Some(lsn) = lsn {
+            self.insert_pending_flush_lsn(lsn);
+        } else {
+            assert!(
+                xact_id.is_some(),
+                "LSN should be none for non streaming flush"
+            );
+        }
+
+        tokio::task::spawn(async move {
+            let mut disk_slice_clone = DiskSliceWriter::new(
+                schema,
+                path,
+                batches,
+                lsn,
+                next_file_id,
+                old_index,
+                parquet_flush_threshold_size,
+            );
+            let flush_result = disk_slice_clone.write().await;
+            match flush_result {
+                Ok(()) => {
+                    table_notify_tx
+                        .send(TableEvent::FlushResult {
+                            xact_id,
+                            flush_result: Some(Ok(disk_slice_clone)),
+                        })
+                        .await
+                        .unwrap();
+                }
+                Err(e) => {
+                    table_notify_tx
```

Comment:
You need to coordinate with drop table, otherwise send would fail.

Reply:
https://github.com/Mooncake-Labs/moonlink/blob/dff3672daad96bca992de37315fd19f529b844ad/src/moonlink/src/table_handler/table_handler_state.rs#L327-L341